src/core/net/dns_client.cpp

```diff
@@ -788,6 +779,11 @@ void Client::ProcessResponse(const Message &aMessage)
 
     response.mMessage = &aMessage;
 
+    // We intentionally parse the response in a separate method
+    // `ParseResponse()` to free all the stack allocated variables
+    // (e.g., `QueryInfo`) used during parsing of the message before
+    // finalizing the query and invoking the user's callback.
```

Comment:
Yes. I noticed same requirement when implementing DNS server code. I tried to make sure that `char name[256]` for storing DNS names are only used in leaf functions. 

Wonder if we can have some tool that can analyze the stack behavior of the OT program.